pylabrobot/resources/ml_star/mfx_modules.py

```diff
@@ -0,0 +1,228 @@
+""" ML Star MFX modules (including machine defintions placed on a MFX carrier) """
+
+from __future__ import annotations
+
+import logging
+from typing import List, Optional, Union
+
+from pylabrobot.resources.resource import Resource
+from pylabrobot.resources.carrier import Coordinate
+
+logger = logging.getLogger("pylabrobot")
+
+
+class MFXSite(Resource):
+  """ A single site within a MFX module. """
+
+  def __init__(self, name: str, size_x: float, size_y: float, size_z: float, spot: int,
+    category: str = "mfx_site", model: Optional[str] = None):
+    super().__init__(name=name, size_x=size_x, size_y=size_y, size_z=size_z, category=category,
+      model=model)
+    self.resource: Optional[Resource] = None
+    self.spot: int = spot
+
+  def assign_child_resource(
+    self,
+    resource: Resource,
+    location: Optional[Coordinate],
+    reassign: bool = True
+  ):
+    self.resource = resource
+    return super().assign_child_resource(resource, location=location)
+
+  def unassign_child_resource(self, resource):
+    self.resource = None
+    return super().unassign_child_resource(resource)
+
+  def serialize(self):
+    return {
+      "spot": self.spot,
+      **super().serialize()
+    }
+
+  def __eq__(self, other):
+    return super().__eq__(other) and self.spot == other.spot and self.resource == other.resource
+
+def create_mfx_module_sites(
+  locations: List[Coordinate],
+  site_size_x: List[Union[float, int]],
+  site_size_y: List[Union[float, int]]) -> List[MFXSite]:
+  """ Create a list of MFX module sites with the given sizes. """
+
+  sites = []
+  for spot, (l, x, y) in enumerate(zip(locations, site_size_x, site_size_y)):
+    site = MFXSite(
+      name = f"mfx-module-site-{spot}",
+      size_x=x, size_y=y, size_z=0, spot=spot)
+    site.location = l
+    sites.append(site)
+  return sites
+
+
+def create_homogeneous_mfx_module_sites(
+  locations: List[Coordinate],
+  site_size_x: float,
+  site_size_y: float) -> List[MFXSite]:
+  """ Create a list of MFX module sites with the same size. """
+
+  n = len(locations)
+  return create_mfx_module_sites(locations, [site_size_x] * n, [site_size_y] * n)
+
+
+class MFXModule(Resource):
+  """ Abstract base resource for MFX modules to be placed
+    on a MFX carrier (landscape/portrait, 4/5 positions).
```

Comment:
This is essentially duplicating `Carrier`. Can we subclass `Carrier` if it's important to have a separate `MFXCarrier`? It is possible the `Carrier` class provides everything you need, in which case I'd discourage adding a new class.

Reply:
You are right, it is essentially duplicating `Carrier`.

But while the software definition is essentially the same, the use is very different and emulates the way that one physically builds MFX_carriers: 
take empty carrier -> choose slot -> screw in MFX_module -> add labware to MFX_module

Continuing with `Carrier` this definition becomes... hard to understand:
take empty `Carrier` -> choose slot -> screw in "small" `Carrier` -> add labware to "small" Carrier's slot?

Hamilton also provides an avalanche of different MFX_modules. Exposing these to PLR will enable complete customisation of Hamilton deck layouts in PLR with no extra time commitment (as opposed to every person having to figure out how to define their MFX_modules in PLR from scratch).